OBLS-328 Initial backend for PickTask resources

druchniewicz opened the pull request.

jmiranda commented on grails-app/services/org/pih/warehouse/picking/PickTaskService.groovy:
```diff
@@ -0,0 +1,71 @@
+
+        Integer ordersCount = command.ordersCount ?: 1
+        String pickType = command.pickType
+        boolean searchForAllOrders = !pickType || pickType == "ALL"
```
The absence of a delivery type code means "system-directed" just to be clear.

jmiranda commented on grails-app/controllers/org/pih/warehouse/api/picking/SearchPickTaskCommand.groovy:
```diff
@@ -0,0 +1,17 @@
+class SearchPickTaskCommand implements Validateable {
+
+    Location facility
+    String pickType
```
To confirm, the pick type was used on the OBKN branch to signify the type of pick being completed (i.e., pallet vs case vs each or piece pick). 

<img width="3116" height="1572" alt="Image" src="https://github.com/user-attachments/assets/1d18fbb0-e54e-4f6b-9ea1-39c235d4bd39" />

druchniewicz replied:
thanks, I changed it do deliveryTypeCode

druchniewicz commented on grails-app/controllers/org/pih/warehouse/api/picking/SearchPickTaskCommand.groovy:
```diff
@@ -0,0 +1,17 @@
+
+    Location facility
+    String pickType
+    Integer ordersCount
```
I wanted set ordersCount to 1 by default, but it bothered me a bit when I wanted to, e.g. get pick tasks for a given user GET /api/pick-tasks?assignee=jmiranda
we would like to fetch all tasks for this user if we do not provide ordersCount param. I left it without a default value for now

jmiranda replied:
Makes sense.

jmiranda commented on grails-app/domain/org/pih/warehouse/picking/PickTask.groovy:
```diff
@@ -0,0 +1,87 @@
+    BigDecimal quantity = 0
+    BigDecimal quantityPicked = 0
+
+    Person pickedBy
```
Yes, I think having the following columns on the picklist item is the way to go in the long run.
* picker_id (assignee_id)
* picked_by_id
* date_assigned
* date_started
* date_picked

jmiranda commented on grails-app/migrations/views/pick-task.sql:
```diff
@@ -0,0 +1,47 @@
+    r.request_number AS requisition_number,
+    r.delivery_type_code AS delivery_type_code,
+    r.date_requested AS date_requested,
+    CASE r.delivery_type_code
```
It's actually two separate priorities. I described it in a ticket somewhere recently. 

The priority associated with the delivery type code is 0-5 with 0 being highest priority. The priority on requisition is 0 to 999 (or even higher) where the 999 is the highest priority. 

To make this more clear, I believe the priority logic will look something like this. 
```
select * 
from requistion
order by r.delivery_type_code.priority asc, r.priority desc
```
This gives us a way to store all requisitions by the delivery type code and then within that by a separate priority. 

Note: Obviously we can't actually run a query like that since the `r.delivery_type_code.priority` is not a thing i.e. delivery type code is just a string (e.g. "WILL_CALL") that maps to an enum DeliveryTypeCode.WILL_CALL. But we could store enums in the database using their "ordinal" value. 

Here's an example of what that looks like in Grails.

<img width="1358" height="1455" alt="image" src="https://github.com/user-attachments/assets/d423bc66-9e5d-4364-a9f2-9a6a884ba496" />

jmiranda replied:
@druchniewicz just checking if you saw this.

druchniewicz replied:
@jmiranda yeah I saw it, thanks for this suggestion but when I'm thinking about this once again I'm not sure if it's the best idea to change deliveryTypeCode from 'string' to 'ordinal':

* now when we have 'string' we can at least explicitly see what type of delivery type it is
* when some delivery types have the same priority (like Local Delivery and Service = 2) it would be difficult to determine what type of task this is, e.g. if you wanted to create a report grouped by delivery code, you wouldn't be able to distinguish between types with the same priority

so I would probably stick with the delivery code being the string

jmiranda replied:
@druchniewicz yeah, i agree with this. and even if the priorities were all unqiue, i can imagine a case where we change the priority and things go awry. 

We'll just need to implement the sort in code or add a hard-coded CASE statement to the SQL query. The latter sounds better since we won't be changing the priority mapping often.

jmiranda commented on grails-app/migrations/views/pick-task.sql:
```diff
@@ -0,0 +1,47 @@
+    r.origin_id AS facility_id,
+    r.type AS requisition_type,
+    r.date_approved as date_started,
+    r.approved_by_id as assignee_id,
```
sounds good. i dumped a draft version of the ideal picklist_item table in a confluence page. i need to compare against the current data model and make sure to include anything i missed on the first go-round. if something is missing it's not because i want it deleted. i just wasn't as thorough because i had limited time.

https://openboxes.atlassian.net/wiki/x/CIAojg

jmiranda replied:
i'll keep dumping table definitions in that document as we go, including new tables like allocation.

jmiranda commented on grails-app/services/org/pih/warehouse/picking/PickTaskService.groovy:
```diff
@@ -0,0 +1,163 @@
+        return task
+    }
+
+    void start(PickTask task, String assigneeId) {
```
I mentioned in OBLS-331 that this request should be executed at the point where the picker is at the physical location, scanning the location barcode. I initially thought it would be earlier, but I couldn't think of an explicit event that made sense (i.e. maybe when we get the pick task sent to the device?)

https://openboxes.atlassian.net/browse/OBLS-331

jmiranda replied:
Feel free to add comments to the ticket if you have thoughts.

druchniewicz commented:
Added next part of code: new fields in PicklistItem (and corresponding fields in PickTask), added API for pick and short-pick  (transfer to container with OUTBOUND_CONTAINER activity code)

druchniewicz commented:
Next part of code has been added. @awalkowiak @jmiranda can you take a look at PR as a whole?

jmiranda commented on grails-app/domain/org/pih/warehouse/requisition/Requisition.groovy:
```diff
@@ -126,6 +126,12 @@ class Requisition implements Comparable<Requisition>, Serializable {
     Boolean approvalRequired
 
+    Date datePicked
+    Person pickedBy
```
These should be on the picklist_item. I believe they should already be there.

jmiranda replied:
<img width="935" height="1290" alt="Image" src="https://github.com/user-attachments/assets/74fc9310-bc77-4fe1-9746-3efcbcaf4b8b" />

jmiranda commented on grails-app/domain/org/pih/warehouse/requisition/Requisition.groovy:
```diff
@@ -126,6 +126,12 @@ class Requisition implements Comparable<Requisition>, Serializable {
+    Person pickedBy
+
+    Date dateStaged
+    Person stagedBy
```
These should also be on the picklist item.

jmiranda replied:
Can you explain the rationale for putting these in the Requisition? Maybe I missed something.

cc @awalkowiak

druchniewicz replied:
> These should also be on the picklist item.

yes, it is on picklist item as well

druchniewicz replied:
> Can you explain the rationale for putting these in the Requisition? Maybe I missed something.
> 
> cc @awalkowiak

I thought about it like following:
We have order e.g. with 3 items (3x pick tasks). We move each of them from outbound container to staging location and change status for each of them to STAGED (on picklist item level). But when we move the whole content of container (3 items) to staging location it means that whole order has been staged - that's why I set it on requisition level as well. Does it make sense?

awalkowiak replied:
I think you don't need it here. If you need it to access it on the requisition, you can derive it from picklist items.

jmiranda commented on grails-app/i18n/messages.properties:
```diff
@@ -981,6 +982,7 @@ enum.RequisitionStatus.REQUESTED=Requested
 enum.RequisitionStatus.PENDING_APPROVAL=Waiting for approval
 enum.RequisitionStatus.APPROVED=Approved
 enum.RequisitionStatus.REJECTED=Rejected
+enum.RequisitionStatus.STAGED=Staged
```
I agree with adding this new state if it's necessary, but be sure to add a ticket to figure out how to handle this in the stock movement workflow. Throwing an exception for unknown statuses seems like the most appropriate (defensive) approach for now.

awalkowiak commented on grails-app/controllers/org/pih/warehouse/api/picking/PickTaskApiController.groovy:
```diff
@@ -0,0 +1,79 @@
+
+        render ([
+                data: tasks,
+                totalCount: tasks.size() ?: 0,
```
I think it should be `tasks.totalCount`, because the `.size()` will be equal to the current "page" of data

jmiranda replied:
Good catch. This one is important to remember. The totalCount is only available when the actual query has been executed through GORM (i.e. HQL, dynamic finder).

Aside: If we ever need to write straight SQL for any queries we should try to abide by the convention of returning results with total count. I can't remember the proper mechanism for that but I believe we have examples in the code (IIRC it requires two queries: one for the results and one for the count).

druchniewicz replied:
indeed, nice finding. Corrected

awalkowiak commented on grails-app/services/org/pih/warehouse/inventory/ProductAvailabilityService.groovy:
```diff
@@ -751,6 +751,45 @@ class ProductAvailabilityService {
+            return []
+        }
+
+        String quantityCondition = excludeZeroQuantity ? "and pa.quantityOnHand > 0" : ""
```
I think we should always exclude zero QoH, not sure if this part is needed

druchniewicz replied:
Corrected

jmiranda commented on src/main/groovy/org/pih/warehouse/api/PickTaskStatus.groovy:
```diff
@@ -8,7 +8,7 @@ enum PickTaskStatus {
     static final Map<PickTaskStatus, Set<PickTaskStatus>> ALLOWED_STATE_TRANSITIONS = [
             (PENDING): [PICKING, PICKED] as Set,
-            (PICKING): [PICKED] as Set,
+            (PICKING): [PICKING, PICKED] as Set,
```
Let's add a comment to explain when a PICKING -> PICKING transition might occur

jmiranda commented on grails-app/controllers/org/pih/warehouse/api/picking/PickUrlMappings.groovy:
```diff
@@ -0,0 +1,21 @@
+            action = [GET: "read", PATCH: "patch"]
+        }
+
+        "/api/facilities/$facility/pick-tasks/containers/$outboundContainerId" {
```
I don't remember if I came up with this, but I'd also be fine with this URL mapping if we start to get some janky looking URLs.
```
/api/facilities/$facility/containers/$outboundContainerId
```

jmiranda replied:
Oh yeah I see the one I was expecting below. I think you can use that one. These containers are likely going to be used with pick tasks but I could see them being used for other activities as well.

druchniewicz commented:
> @druchniewicz overall it looks good to me, there are two really minor things

@awalkowiak thanks, changes have been applied

jmiranda commented on src/main/groovy/org/pih/warehouse/requisition/RequisitionStatus.groovy:
```diff
@@ -26,10 +26,11 @@ enum RequisitionStatus {
-    CANCELED(13, null, StatusType.DANGER),
-    DELETED(14, null, StatusType.DANGER),
-    ERROR(15, null, StatusType.DANGER),
+    STAGED(12, null, StatusType.SUCCESS),
```
Staged actually comes before Issued. Sorry for not clarifying that.

awalkowiak commented on grails-app/controllers/org/pih/warehouse/UrlMappings.groovy:
```diff
@@ -1139,6 +1139,11 @@ class UrlMappings {
             action = [POST: "refreshCycleCount"]
         }
 
+        "/api/facilities/$facility/containers/${outboundContainerId}" {
```
I think this is a redundant endpoint. Looks like there is no `patch` action in the `LocationApiController`
